scripts/safe_template_linter.py

```diff
@@ -1365,10 +1365,32 @@ def _check_concat_with_html(self, file_contents, rule, results):
         """
         lines = StringLines(file_contents)
         last_expression = None
-        # attempt to match a string that starts with '<' or ends with '>'
-        regex_string_with_html = r"""["'](?:\s*<.*|.*>\s*)["']"""
-        regex_concat_with_html = r"(\+\s*{}|{}\s*\+)".format(regex_string_with_html, regex_string_with_html)
-        for match in re.finditer(regex_concat_with_html, file_contents):
+        # Match quoted strings that starts with '<' or ends with '>'.
+        regex_string_with_html = r"""
+            {quote}                             # Opening quote.
+                (
+                   \s*<                         # Starts with '<' (ignoring spaces)
+                   ([^{quote}]|[\\]{quote})*    # followed by anything but a closing quote.
+                |                               # Or,
+                   ([^{quote}]|[\\]{quote})*    # Anything but a closing quote
+                   >\s*                         # ending with '>' (ignoring spaces)
+                )
+            {quote}                             # Closing quote.
+        """
+        # Match single or double quote.
+        regex_string_with_html = "({}|{})".format(
+            regex_string_with_html.format(quote="'"),
+            regex_string_with_html.format(quote='"'),
+        )
+        # Match quoted HTML strings next to a '+'.
+        regex_concat_with_html = re.compile(
+            r"(\+\s*{}|{}\s*\+)".format(
+                regex_string_with_html.format(quote='quote1'),
+                regex_string_with_html.format(quote='quote2')
```

Comment:
My point is that if we break this into three steps (find string literals, filter those to strings with angle brackets at the end, filter those that are next to plusses), each step will be easier to understand, test, and debug.  I don't want to see the next step in this regex's evolution!